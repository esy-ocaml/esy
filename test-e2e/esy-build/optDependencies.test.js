// @flow

const helpers = require('../test/helpers');

function makePackage(p, {name, build, dependencies, optDependencies}, ...children) {
  return [
    helpers.packageJson({
      name: name,
      version: '1.0.0',
      esy: {
        build,
      },
      dependencies,
      optDependencies,
    }),
    ...children,
  ];
}

describe('Build with optDependencies', () => {
  it('builds w/o opt dependency installed', async () => {
    const p = await helpers.createTestSandbox();

    await p.fixture(
      ...makePackage(
        p,
        {
          name: 'root',
          build: 'true',
          dependencies: {
            dep: 'path:./dep',
          },
          optDependencies: {},
        },
        helpers.dir(
          'dep',
          ...makePackage(p, {
            name: 'dep',
            build: 'optDep.installed #{optDep.installed}',
            dependencies: {},
            optDependencies: {optDep: '*'},
          }),
        ),
      ),
    );

    await p.esy('install');
    const plan = JSON.parse((await p.esy('build-plan dep@path:dep')).stdout);
    expect(plan.build).toEqual([['optDep.installed', 'false']]);

    {
      const {stdout} = await p.esy('build-env');
      expect(p.normalizePathsForSnapshot(stdout)).toMatchSnapshot();
    }

    {
      const {stdout} = await p.esy('build-env dep');
      expect(p.normalizePathsForSnapshot(stdout)).toMatchSnapshot();
    }
  });

  it('builds w/ opt dependency installed', async () => {
    const p = await helpers.createTestSandbox();

    await p.fixture(
      ...makePackage(
        p,
        {
          name: 'root',
          build: 'true',
          dependencies: {
            dep: 'path:./dep',
            optDep: 'path:./optDep',
          },
          optDependencies: {},
        },
        helpers.dir(
          'dep',
          ...makePackage(p, {
            name: 'dep',
            build: 'optDep.installed #{optDep.installed}',
            dependencies: {},
            optDependencies: {optDep: '*'},
          }),
        ),
        helpers.dir(
          'optDep',
          ...makePackage(p, {
            name: 'optDep',
            build: 'true',
            dependencies: {},
            optDependencies: {},
          }),
        ),
      ),
    );

    await p.esy('install');
    const plan = JSON.parse((await p.esy('build-plan dep@path:dep')).stdout);
    expect(plan.build).toEqual([['optDep.installed', 'true']]);

    {
      const {stdout} = await p.esy('build-env');
      expect(p.normalizePathsForSnapshot(stdout)).toMatchSnapshot();
    }

    {
      const {stdout} = await p.esy('build-env dep');
      expect(p.normalizePathsForSnapshot(stdout)).toMatchSnapshot();
    }
  });

  it('opam package builds w/ opt dependency installed', async () => {
    const p = await helpers.createTestSandbox();

    await p.defineNpmPackage({
      name: '@esy-ocaml/substs',
      version: '1.0.0',
      esy: {},
    });

    await p.defineOpamPackage({
      name: 'dep',
      version: '1.0.0',
      opam: `
      opam-version: "2.0"
      build: [
        ["optDep:installed" "%{optDep:installed}%"]
        ["optDep:enabled" "%{optDep:enabled}%"]
      ]
      depopts: ["optDep"]
      `,
      url: null,
    });

    await p.defineOpamPackage({
      name: 'optDep',
      version: '1.0.0',
      opam: `
      opam-version: "2.0"
      build: [
        "true"
      ]
      `,
      url: null,
    });

    await p.fixture(
      ...makePackage(p, {
        name: 'root',
        build: 'true',
        dependencies: {
          '@opam/dep': '*',
          '@opam/optDep': '*',
        },
        optDependencies: {},
      }),
    );

    await p.esy('install');
    const plan = JSON.parse((await p.esy('build-plan @opam/dep@opam:1.0.0')).stdout);
    expect(plan.build).toEqual([
      ['optDep:installed', 'true'],
      ['optDep:enabled', 'enable'],
    ]);
  });

  it('opam package builds w/o opt dependency installed', async () => {
    const p = await helpers.createTestSandbox();

    await p.defineNpmPackage({
      name: '@esy-ocaml/substs',
      version: '1.0.0',
      esy: {},
    });

    await p.defineOpamPackage({
      name: 'dep',
      version: '1.0.0',
      opam: `
      opam-version: "2.0"
      build: [
        ["optDep:installed" "%{optDep:installed}%"]
        ["optDep:enabled" "%{optDep:enabled}%"]
      ]
      depopts: ["optDep"]
      `,
      url: null,
    });

    await p.defineOpamPackage({
      name: 'optDep',
      version: '1.0.0',
      opam: `
      opam-version: "2.0"
      build: [
        "true"
      ]
      `,
      url: null,
    });

    await p.fixture(
      ...makePackage(p, {
        name: 'root',
        build: 'true',
        dependencies: {
          '@opam/dep': '*',
        },
        optDependencies: {},
      }),
    );

    await p.esy('install');
    const plan = JSON.parse((await p.esy('build-plan @opam/dep@opam:1.0.0')).stdout);
    expect(plan.build).toEqual([
      ['optDep:installed', 'false'],
      ['optDep:enabled', 'disable'],
    ]);
  });
});
