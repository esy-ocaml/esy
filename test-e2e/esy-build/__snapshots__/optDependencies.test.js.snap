// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build with optDependencies builds w/ opt dependency installed 1`] = `
"# Build environment for root@link:./package.json

#
# optDep@1.0.0
#
export CAML_LD_LIBRARY_PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/optdep-f9d4268f/stublibs:%esyPrefixPath%/3_________________________________________________________/i/optdep-f9d4268f/lib/stublibs:$CAML_LD_LIBRARY_PATH\\"
export OCAMLPATH=\\"%esyPrefixPath%/3_________________________________________________________/i/optdep-f9d4268f/lib:$OCAMLPATH\\"
export MAN_PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/optdep-f9d4268f/man:$MAN_PATH\\"
export PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/optdep-f9d4268f/bin:$PATH\\"

#
# dep@1.0.0
#
export CAML_LD_LIBRARY_PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/dep-09cf1d25/stublibs:%esyPrefixPath%/3_________________________________________________________/i/dep-09cf1d25/lib/stublibs:$CAML_LD_LIBRARY_PATH\\"
export OCAMLPATH=\\"%esyPrefixPath%/3_________________________________________________________/i/dep-09cf1d25/lib:$OCAMLPATH\\"
export MAN_PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/dep-09cf1d25/man:$MAN_PATH\\"
export PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/dep-09cf1d25/bin:$PATH\\"

#
# root@1.0.0
#
export OCAMLFIND_LDCONF=\\"ignore\\"
export OCAMLFIND_DESTDIR=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb/lib\\"
export cur__etc=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb/etc\\"
export cur__share=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb/share\\"
export cur__toplevel=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb/toplevel\\"
export cur__stublibs=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb/stublibs\\"
export cur__doc=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb/doc\\"
export cur__man=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb/man\\"
export cur__lib=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb/lib\\"
export cur__sbin=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb/sbin\\"
export cur__bin=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb/bin\\"
export cur__install=\\"%projectPath%/_esy/default/store/i/root-fcc61fbb\\"
export cur__target_dir=\\"%projectPath%/_esy/default/store/b/root-fcc61fbb\\"
export cur__original_root=\\"%projectPath%\\"
export cur__root=\\"%projectPath%\\"
export cur__dev=\\"true\\"
export cur__version=\\"1.0.0\\"
export cur__name=\\"root\\"
export DUNE_BUILD_DIR=\\"%projectPath%/_esy/default/store/b/root-fcc61fbb\\"

#
# Built-in
#
export SHELL=\\"env -i /bin/bash --norc --noprofile\\"
export PATH=\\"$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin\\""
`;

exports[`Build with optDependencies builds w/ opt dependency installed 2`] = `
"# Build environment for dep@path:dep

#
# optDep@1.0.0
#
export CAML_LD_LIBRARY_PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/optdep-f9d4268f/stublibs:%esyPrefixPath%/3_________________________________________________________/i/optdep-f9d4268f/lib/stublibs:$CAML_LD_LIBRARY_PATH\\"
export OCAMLPATH=\\"%esyPrefixPath%/3_________________________________________________________/i/optdep-f9d4268f/lib:$OCAMLPATH\\"
export MAN_PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/optdep-f9d4268f/man:$MAN_PATH\\"
export PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/optdep-f9d4268f/bin:$PATH\\"

#
# dep@1.0.0
#
export OCAMLFIND_LDCONF=\\"ignore\\"
export OCAMLFIND_DESTDIR=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25/lib\\"
export cur__etc=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25/etc\\"
export cur__share=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25/share\\"
export cur__toplevel=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25/toplevel\\"
export cur__stublibs=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25/stublibs\\"
export cur__doc=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25/doc\\"
export cur__man=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25/man\\"
export cur__lib=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25/lib\\"
export cur__sbin=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25/sbin\\"
export cur__bin=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25/bin\\"
export cur__install=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-09cf1d25\\"
export cur__target_dir=\\"%esyPrefixPath%/3_________________________________________________________/b/dep-09cf1d25\\"
export cur__original_root=\\"%esyPrefixPath%/esyi/source/i/dep--67b360d9a56facfd6fc9b90817f589ec\\"
export cur__root=\\"%esyPrefixPath%/esyi/source/i/dep--67b360d9a56facfd6fc9b90817f589ec\\"
export cur__dev=\\"false\\"
export cur__version=\\"1.0.0\\"
export cur__name=\\"dep\\"
export DUNE_BUILD_DIR=\\"%esyPrefixPath%/3_________________________________________________________/b/dep-09cf1d25\\"

#
# Built-in
#
export SHELL=\\"env -i /bin/bash --norc --noprofile\\"
export PATH=\\"$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin\\""
`;

exports[`Build with optDependencies builds w/o opt dependency installed 1`] = `
"# Build environment for root@link:./package.json

#
# dep@1.0.0
#
export CAML_LD_LIBRARY_PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/dep-53a9a3ba/stublibs:%esyPrefixPath%/3_________________________________________________________/i/dep-53a9a3ba/lib/stublibs:$CAML_LD_LIBRARY_PATH\\"
export OCAMLPATH=\\"%esyPrefixPath%/3_________________________________________________________/i/dep-53a9a3ba/lib:$OCAMLPATH\\"
export MAN_PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/dep-53a9a3ba/man:$MAN_PATH\\"
export PATH=\\"%esyPrefixPath%/3_________________________________________________________/i/dep-53a9a3ba/bin:$PATH\\"

#
# root@1.0.0
#
export OCAMLFIND_LDCONF=\\"ignore\\"
export OCAMLFIND_DESTDIR=\\"%projectPath%/_esy/default/store/i/root-80f266a4/lib\\"
export cur__etc=\\"%projectPath%/_esy/default/store/i/root-80f266a4/etc\\"
export cur__share=\\"%projectPath%/_esy/default/store/i/root-80f266a4/share\\"
export cur__toplevel=\\"%projectPath%/_esy/default/store/i/root-80f266a4/toplevel\\"
export cur__stublibs=\\"%projectPath%/_esy/default/store/i/root-80f266a4/stublibs\\"
export cur__doc=\\"%projectPath%/_esy/default/store/i/root-80f266a4/doc\\"
export cur__man=\\"%projectPath%/_esy/default/store/i/root-80f266a4/man\\"
export cur__lib=\\"%projectPath%/_esy/default/store/i/root-80f266a4/lib\\"
export cur__sbin=\\"%projectPath%/_esy/default/store/i/root-80f266a4/sbin\\"
export cur__bin=\\"%projectPath%/_esy/default/store/i/root-80f266a4/bin\\"
export cur__install=\\"%projectPath%/_esy/default/store/i/root-80f266a4\\"
export cur__target_dir=\\"%projectPath%/_esy/default/store/b/root-80f266a4\\"
export cur__original_root=\\"%projectPath%\\"
export cur__root=\\"%projectPath%\\"
export cur__dev=\\"true\\"
export cur__version=\\"1.0.0\\"
export cur__name=\\"root\\"
export DUNE_BUILD_DIR=\\"%projectPath%/_esy/default/store/b/root-80f266a4\\"

#
# Built-in
#
export SHELL=\\"env -i /bin/bash --norc --noprofile\\"
export PATH=\\"$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin\\""
`;

exports[`Build with optDependencies builds w/o opt dependency installed 2`] = `
"# Build environment for dep@path:dep

#
# dep@1.0.0
#
export OCAMLFIND_LDCONF=\\"ignore\\"
export OCAMLFIND_DESTDIR=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba/lib\\"
export cur__etc=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba/etc\\"
export cur__share=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba/share\\"
export cur__toplevel=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba/toplevel\\"
export cur__stublibs=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba/stublibs\\"
export cur__doc=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba/doc\\"
export cur__man=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba/man\\"
export cur__lib=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba/lib\\"
export cur__sbin=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba/sbin\\"
export cur__bin=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba/bin\\"
export cur__install=\\"%esyPrefixPath%/3_________________________________________________________/s/dep-53a9a3ba\\"
export cur__target_dir=\\"%esyPrefixPath%/3_________________________________________________________/b/dep-53a9a3ba\\"
export cur__original_root=\\"%esyPrefixPath%/esyi/source/i/dep--67b360d9a56facfd6fc9b90817f589ec\\"
export cur__root=\\"%esyPrefixPath%/esyi/source/i/dep--67b360d9a56facfd6fc9b90817f589ec\\"
export cur__dev=\\"false\\"
export cur__version=\\"1.0.0\\"
export cur__name=\\"dep\\"
export DUNE_BUILD_DIR=\\"%esyPrefixPath%/3_________________________________________________________/b/dep-53a9a3ba\\"

#
# Built-in
#
export SHELL=\\"env -i /bin/bash --norc --noprofile\\"
export PATH=\\"$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin\\""
`;
